Chapter level specification

Enkidu93 opened the pull request:
Up for name ideas for what is now called `BiblicalRange` - trying to avoid something too wordy or unclear.

Addresses #150 .

Enkidu93 commented:
I'm realizing I should add one more test to cover the error when both are specified. I'll do that.

ddaspit reviewed (changes requested):
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 2 unresolved discussions (waiting on @Enkidu93 and @johnml1135)

___
*[`src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs` line 12 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5me7d1ZvOXg1Y2gxb:-Np5me7d1ZvOXg1Y2gxc:b-ynw4i) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs#L12)):*
> ```Smalltalk
>     public IList<string>? TextIds { get; set; }
> 
>     public string? BiblicalRange { get; set; }
> ```

I would rename this to `ScriptureRange` along with all other related properties. 
___
*[`tests/Serval.E2ETests/ServalApiTests.cs` line 395 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5nUwB5k4FrNVNCDMS:-Np5nUwC5rWo7eaHr00s:bz1rjal) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/tests/Serval.E2ETests/ServalApiTests.cs#L395)):*
> ```Smalltalk
>             (await _helperClient.translationEnginesClient.GetAllBuildsAsync(engineId)).First().State
>                 == JobState.Completed,
>             JsonSerializer.Serialize((await _helperClient.translationEnginesClient.GetAllBuildsAsync(engineId)).First())
> ```

It is unclear to me what this is asserting.


<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`tests/Serval.E2ETests/ServalApiTests.cs` line 395 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5nUwB5k4FrNVNCDMS:-Np5q7NxBysg_ZlDzsJe:b-t5r9ti) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/tests/Serval.E2ETests/ServalApiTests.cs#L395)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

It is unclear to me what this is asserting.
</blockquote></details>

The only thing I changed was that I added an error message. It's declaring that the (one and only) build is complete. I could refactor the helperClient's `BuildEngine()` to return the specific id of the build and get it by id if you'd prefer.


<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs` line 12 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5me7d1ZvOXg1Y2gxb:-Np5pjr9DgXKdrSv3Qde:b-wi0g9w) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs#L12)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I would rename this to `ScriptureRange` along with all other related properties.
</blockquote></details>

OK, just because we use 'scripture' more elsewhere in the code? 


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Grpc/Protos/serval/translation/v1/engine.proto` line 139 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5sFofEveDZPWqtXW2:-Np5sFofEveDZPWqtXW3:b2dbqvb) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Grpc/Protos/serval/translation/v1/engine.proto#L139)):*
> ```proto
>     bool pretranslate_all = 5;
>     string train_on_biblical_range = 6;
>     string pretranslate_biblical_range = 7;
> ```

Should these be optional?


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (commented):
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 3 unresolved discussions (waiting on @ddaspit and @Enkidu93)



<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Grpc/Protos/serval/translation/v1/engine.proto` line 139 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5sFofEveDZPWqtXW2:-Npfigsl5tjn9IpVlUbC:b-235ihu) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Grpc/Protos/serval/translation/v1/engine.proto#L139)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Should these be optional?
</blockquote></details>

Will all chapters be put here, or only ones that have chapters?


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 986 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-NpflDLJ1PxX-S1b1Nek:-NpflDLJ1PxX-S1b1Nel:bc9x8lu) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L986)):*
> ```Smalltalk
>                         CorpusRef = ptcc.CorpusId,
>                         TextIds = ptcc.TextIds?.ToList(),
>                         Chapters = ptcc.Chapters
> ```

This is a build error - your code doesn't build :-(.  There are a few of these throughout.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Services/EngineService.cs` line 196 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-NpfnAdl4UrJ18JPVqJs:-NpfnAdl4UrJ18JPVqJt:b-im4m2e) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Services/EngineService.cs#L196)):*
> ```Smalltalk
>                                             (kvp) =>
>                                             {
>                                                 var scriptureChapters = new ScriptureChapters();
> ```

Can you explain what this code exactly is doing?  Also, why is it duplicated below?  Could it be put into a separate function?


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-Npfjt498SQ4oXVueDZO:-Npfjt498SQ4oXVueDZP:b2jk154) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
> ```Smalltalk
>     public IList<string>? TextIds { get; set; }
> 
>     public string? Chapters { get; set; }
> ```

Shouldn't this be a ScriptureRange?  It is books and chapters...  There are also build errors in this code.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Contracts/TrainingCorpusDto.cs` line 9 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-Npfk-Pf1O2KuTj7GJUe:-Npfk-Pf1O2KuTj7GJUf:bql51rz) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Contracts/TrainingCorpusDto.cs#L9)):*
> ```Smalltalk
>     public IList<string>? TextIds { get; set; }
> 
>     public string? Chapters { get; set; }
> ```

Same as other question...


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (commented):
> Reviewed 11 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 7 unresolved discussions (waiting on @ddaspit and @Enkidu93)



<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 967 at r3](https://reviewable.io/reviews/sillsdev/serval/289#-NpjAiuk61KciZSu2f0m:-NpjAiuk61KciZSu2f0n:bjr8296) ([raw file](https://github.com/sillsdev/serval/blob/635b0abd01fe5360b33afcdfd622573ff9984c24/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L967)):*
> ```Smalltalk
>         var build = new Build { EngineRef = engine.Id, Name = source.Name };
>         var corpusIds = new HashSet<string>(engine.Corpora.Select(c => c.Id));
>         ScriptureRangeParser parser = new();
> ```

This line causes my build to fail.


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (changes requested):
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 6 unresolved discussions (waiting on @ddaspit and @Enkidu93)

___
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-Npfjt498SQ4oXVueDZO:-NpjAXxIDFLDI9TJtlWp:b3r1g0y) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Shouldn't this be a ScriptureRange?  It is books and chapters...  There are also build errors in this code.
</blockquote></details>

Ok - Chapters are the parse parameter (ok).


<!-- Sent from Reviewable.io -->

Enkidu93 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: 11 of 12 files reviewed, 6 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`src/Serval.Grpc/Protos/serval/translation/v1/engine.proto` line 139 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5sFofEveDZPWqtXW2:-NpuvtHn58iirj30tohd:by9m1vp) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Grpc/Protos/serval/translation/v1/engine.proto#L139)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Will all chapters be put here, or only ones that have chapters?
</blockquote></details>

Only ones that have chapters since we're keeping the idiom of 'empty list = all chapters'
___
*[`src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs` line 12 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5me7d1ZvOXg1Y2gxb:-Npuw7QN9pSkRZWl91C4:b-896fix) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs#L12)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

OK, just because we use 'scripture' more elsewhere in the code?
</blockquote></details>

Done.
___
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-Npfjt498SQ4oXVueDZO:-Npuw9yAC51UNVVIEdAl:byr4flj) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Ok - Chapters are the parse parameter (ok).
</blockquote></details>

Right.
___
*[`src/Serval.Translation/Contracts/TrainingCorpusDto.cs` line 9 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-Npfk-Pf1O2KuTj7GJUe:-NpuwCHi-w1rhw-kyZjz:buhag11) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Contracts/TrainingCorpusDto.cs#L9)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Same as other question...
</blockquote></details>

As you said, I wanted to differentiate the string from the post-parsing value.
___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 986 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-NpflDLJ1PxX-S1b1Nek:-NpuwZMo0oZhmd6sPBCB:b-tbuzfl) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L986)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

This is a build error - your code doesn't build :-(.  There are a few of these throughout.
</blockquote></details>

Let me know if any of these errors still exist. I didn't really mean for that commit to be reviewed since it wasn't complete. I just wanted to commit what I had so I could more easily turn to moving the parallel corpus logic. I think I just accidentally pushed.
___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 967 at r3](https://reviewable.io/reviews/sillsdev/serval/289#-NpjAiuk61KciZSu2f0m:-NpuwIQe5NbkleZCtkXE:bi8efxx) ([raw file](https://github.com/sillsdev/serval/blob/635b0abd01fe5360b33afcdfd622573ff9984c24/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L967)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

This line causes my build to fail.
</blockquote></details>

That's because you need the corresponding version of machine. I wonder if we should alter the (non-E2E) CI to do the same thing as the E2E tests (i.e., grab a matching branch of machine if it exists). 
___
*[`src/Serval.Translation/Services/EngineService.cs` line 196 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-NpfnAdl4UrJ18JPVqJs:-Npux63d5JFF44ilAq73:bwz4c6c) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Services/EngineService.cs#L196)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Can you explain what this code exactly is doing?  Also, why is it duplicated below?  Could it be put into a separate function?
</blockquote></details>

Because you can't have a type of `map<string, *repeated* int32>` in proto, I had to create a separate data structure `ScriptureChapters` which is just a placeholder for a list of ints. All this code is doing is mapping a list of ints to a `ScriptureChapters` object. I can extract it into a map function if you like. 
___
*[`tests/Serval.E2ETests/ServalApiTests.cs` line 395 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5nUwB5k4FrNVNCDMS:-NpuxVdi7Yo344-nFRCI:bl4wj53) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/tests/Serval.E2ETests/ServalApiTests.cs#L395)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

The only thing I changed was that I added an error message. It's declaring that the (one and only) build is complete. I could refactor the helperClient's `BuildEngine()` to return the specific id of the build and get it by id if you'd prefer.
</blockquote></details>

Did you want a change here?


<!-- Sent from Reviewable.io -->

ddaspit reviewed (changes requested):
> Reviewed 6 of 11 files at r2, 5 of 6 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 7 unresolved discussions (waiting on @Enkidu93 and @johnml1135)

___
*[`src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs` line 12 at r1](https://reviewable.io/reviews/sillsdev/serval/289#-Np5me7d1ZvOXg1Y2gxb:-Np5qu9J-Y40ur4Taq_j:b-jekbut) ([raw file](https://github.com/sillsdev/serval/blob/d6a5b3985447dfa9108fb6ceb3b71efa44424999/src/Serval.Translation/Contracts/PretranslateCorpusConfigDto.cs#L12)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

OK, just because we use 'scripture' more elsewhere in the code?
</blockquote></details>

Yes, Scripture is the more commonly used term in our codebase.
___
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-NpuxZ6V7gvbfovELtkS:-NpuxZ6V7gvbfovELtkT:b-3ar18p) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
> ```Smalltalk
>     public IList<string>? TextIds { get; set; }
> 
>     public Dictionary<string, List<int>>? Chapters { get; set; }
> ```

It would be better if this was the string that was passed it originally.
___
*[`src/Serval.Translation/Contracts/TrainingCorpusDto.cs` line 9 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-NpuxicI46sgKVZRwyiC:-NpuxicI46sgKVZRwyiD:b-3ar18p) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Contracts/TrainingCorpusDto.cs#L9)):*
> ```Smalltalk
>     public IList<string>? TextIds { get; set; }
> 
>     public Dictionary<string, List<int>>? Chapters { get; set; }
> ```

It would be better if this was the string that was passed it originally.


<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Npv0UAJEQUBLr5P47pE:b-w4k9am) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
> ```Smalltalk
>                         new(
>                             _scriptureDataFileService
>                                 .GetParatextProjectSettings(corpus.SourceFiles.First().Filename)
> ```

Just wondering here: Should we be using source or target versification?


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Services/EngineService.cs` line 196 at r2](https://reviewable.io/reviews/sillsdev/serval/289#-NpfnAdl4UrJ18JPVqJs:-Npv5it3B2nT5RusYlUl:baq5xsf) ([raw file](https://github.com/sillsdev/serval/blob/08976c669c29cccae03d8c03d7e31872612398f2/src/Serval.Translation/Services/EngineService.cs#L196)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

Because you can't have a type of `map<string, *repeated* int32>` in proto, I had to create a separate data structure `ScriptureChapters` which is just a placeholder for a list of ints. All this code is doing is mapping a list of ints to a `ScriptureChapters` object. I can extract it into a map function if you like.
</blockquote></details>

Just wanted to know there was logic :-)


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 4 unresolved discussions (waiting on @ddaspit and @Enkidu93)



<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-NpuxZ6V7gvbfovELtkS:-NpvUJxB-10vaC6JB-qk:bunan8w) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

It would be better if this was the string that was passed it originally.
</blockquote></details>

OK! In order to show the string, I'm going to have to keep track of it along the way parallel to the parsed dictionary Serval-side. What about if we return both? Part of the reason I went for just the dictionary here is that I thought it might be useful because of its clarity and consistency (i.e., there are many strings that will parse to the same chapters output). If a client requests the build information, they/it can verify based on the parsed chapters that things are behaving as expected. Note to self: If we do keep it, I should change the <example>...</example> to make it clear that the keys are books. 


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> Reviewed 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 2 unresolved discussions (waiting on @Enkidu93)

___
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-NpuxZ6V7gvbfovELtkS:-NpyzDW_4HM9iSo5DNwm:b-vb7gun) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

OK! In order to show the string, I'm going to have to keep track of it along the way parallel to the parsed dictionary Serval-side. What about if we return both? Part of the reason I went for just the dictionary here is that I thought it might be useful because of its clarity and consistency (i.e., there are many strings that will parse to the same chapters output). If a client requests the build information, they/it can verify based on the parsed chapters that things are behaving as expected. Note to self: If we do keep it, I should change the <example>...</example> to make it clear that the keys are books.
</blockquote></details>

You don't need to keep track of both. Just parse the string in `EngineService.StartBuildAsync()`. Including the chapters data structure might provide some helpful information, but I wouldn't worry about it right now. We can always add it later on.
___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Npz-F8e8V4bva-PMpgY:b-4ux7xq) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

Just wondering here: Should we be using source or target versification?
</blockquote></details>

That is a good question. How would most people think of it? Maybe we should ask the projects team.


<!-- Sent from Reviewable.io -->

Enkidu93 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: 8 of 13 files reviewed, 2 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`src/Serval.Translation/Contracts/PretranslateCorpusDto.cs` line 9 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-NpuxZ6V7gvbfovELtkS:-NpzF-K60CguSymdUt0n:b-896fix) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Contracts/PretranslateCorpusDto.cs#L9)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

You don't need to keep track of both. Just parse the string in `EngineService.StartBuildAsync()`. Including the chapters data structure might provide some helpful information, but I wouldn't worry about it right now. We can always add it later on.
</blockquote></details>

Done.
___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Npz8ciV-gZK25YpzYMV:bsyrbfn) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

That is a good question. How would most people think of it? Maybe we should ask the projects team.
</blockquote></details>

They agree that the source versification makes the most sense. 


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> Reviewed 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 1 unresolved discussion (waiting on @Enkidu93)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-NpzMEIT-Q6qGx60rP3J:b-76bi0r) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

They agree that the source versification makes the most sense.
</blockquote></details>

That's funny, because we just came to the opposite conclusion in the standup meeting. We thought that the translation team would be primarily focused on the target translation, since that is what they are working on. For pretranslations, they would think: "I want to fill in this blank chapter in my translation." On the other hand, the main userbase for Serval are application developers, so maybe we should think of it from an API perspective and focus on how the app developer would think of it. They might think: "I want to translate these chapters in the source translation."


<!-- Sent from Reviewable.io -->

Enkidu93 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 1 unresolved discussion (waiting on @ddaspit)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-NpzfpKE5xAvBDI5t8h0:b-b7d997) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

That's funny, because we just came to the opposite conclusion in the standup meeting. We thought that the translation team would be primarily focused on the target translation, since that is what they are working on. For pretranslations, they would think: "I want to fill in this blank chapter in my translation." On the other hand, the main userbase for Serval are application developers, so maybe we should think of it from an API perspective and focus on how the app developer would think of it. They might think: "I want to translate these chapters in the source translation."
</blockquote></details>

Oh, haha, yeah, I think the thought was akin to what you mentioned. On a related note, Michael, however, was under the impression that the parallel corpora were being normalized to original versification which isn't the case, right? (If so, where is it in the code :laughing: ?). 


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 1 unresolved discussion (waiting on @Enkidu93)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-NpzhDJe3rJNRVOj1Y4j:b-lqafc6) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

Oh, haha, yeah, I think the thought was akin to what you mentioned. On a related note, Michael, however, was under the impression that the parallel corpora were being normalized to original versification which isn't the case, right? (If so, where is it in the code :laughing: ?).
</blockquote></details>

No, we aren't normalizing to Original versification. Regarding source versification vs. target versification, I would lean towards using the target versification. What was the reasoning that the projects team gave?


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 976 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-NpzhyEN9USLLgYPaIAm:-NpzhyEN9USLLgYPaIAn:bqdytai) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L976)):*
> ```Smalltalk
>                         $"The corpus {ptcc.CorpusId} is not valid: This corpus does not exist for engine {engine.Id}."
>                     );
>                 if (ptcc.TextIds != null && ptcc.ScriptureRange != null)
> ```

If people don't have textId's or Scripture range, we should just select everything, not drop out.


<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Npzit823CGcTsKXEkkF:b-krz9te) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

No, we aren't normalizing to Original versification. Regarding source versification vs. target versification, I would lean towards using the target versification. What was the reasoning that the projects team gave?
</blockquote></details>

OK, excellent: That's what I thought. I'm pinging them again. My impression was that they were thinking similar to what you said above: I want to translate these source chapters.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Npzkd-ADhRNOThDkT87:bjtcyls) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

OK, excellent: That's what I thought. I'm pinging them again. My impression was that they were thinking similar to what you said above: I want to translate these source chapters.
</blockquote></details>

Whatever we decide, we need to put into the swagger documentation.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Services/EngineService.cs` line 188 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-Npzl7WsBkH7hFmrB9za:-Npzl7WsBkH7hFmrB9zb:b-uxemt5) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Services/EngineService.cs#L188)):*
> ```Smalltalk
>                         if (pretranslate?.TryGetValue(c.Id, out PretranslateCorpus? pretranslateCorpus) ?? false)
>                         {
>                             corpus.PretranslateAll =
> ```

This really feels like it could be a method called twice, even if it is defined in this method directly.


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (changes requested):
> Reviewed 3 of 6 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 3 unresolved discussions (waiting on @ddaspit and @Enkidu93)



<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 3 unresolved discussions (waiting on @Enkidu93 and @johnml1135)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 976 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-NpzhyEN9USLLgYPaIAm:-NpznrYSCCErf9z09XO1:bpuj8o) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L976)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

If people don't have textId's or Scripture range, we should just select everything, not drop out.
</blockquote></details>

This exception is thrown if both the text ids and scripture ranges are specified.


<!-- Sent from Reviewable.io -->

Enkidu93 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: 10 of 13 files reviewed, 2 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Nq-JETu-CTV2PrujJx3:bq63nxp) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Whatever we decide, we need to put into the swagger documentation.
</blockquote></details>

Done. Still working on writing up the wiki but the link is there and a brief mention.
___
*[`src/Serval.Translation/Services/EngineService.cs` line 188 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-Npzl7WsBkH7hFmrB9za:-Nq-HObI7I17cTOEHR6n:b-896fix) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Services/EngineService.cs#L188)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

This really feels like it could be a method called twice, even if it is defined in this method directly.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: 10 of 13 files reviewed, 2 unresolved discussions (waiting on @johnml1135)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Nq-QQRhEvXGeT68ld36:b-njf7v1) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, Enkidu93 (Eli C. Lowry) wrote…</i></summary><blockquote>

Done. Still working on writing up the wiki but the link is there and a brief mention.
</blockquote></details>

We talked about this in our AI Drafting meeting this afternoon. We decided to go with the target versification. This would potentially avoid issues in the future if we support multiple source projects in a parallel corpus.


<!-- Sent from Reviewable.io -->

Enkidu93 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 995 at r4](https://reviewable.io/reviews/sillsdev/serval/289#-Npv0UAJEQUBLr5P47pD:-Nq-V8lxAEsfdn2IoNe_:b-4w88o6) ([raw file](https://github.com/sillsdev/serval/blob/ce63d29a65ae6fb75f408f2e051dc19f2c3c8b83/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L995)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

We talked about this in our AI Drafting meeting this afternoon. We decided to go with the target versification. This would potentially avoid issues in the future if we support multiple source projects in a parallel corpus.
</blockquote></details>

Yes, John let me know. That makes sense. Thank you! As you can see, I've already pushed the change to move to target


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 3 unresolved discussions (waiting on @Enkidu93 and @johnml1135)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 985 at r7](https://reviewable.io/reviews/sillsdev/serval/289#-Nq-dVUQDLBh4XGqOja1:-Nq-dVUQDLBh4XGqOja2:b-vsoo0g) ([raw file](https://github.com/sillsdev/serval/blob/59aebce7036937f587a85908f958227cf596f869/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L985)):*
> ```Smalltalk
>                 {
>                     Corpus corpus = engine.Corpora.Where(c => c.Id == ptcc.CorpusId).First();
>                     if (corpus.TargetFiles.Count() > 1 || corpus.TargetFiles.First().Format != FileFormat.Paratext)
> ```

This check should be performed in the `EngineService`. I think I would rather the engine service determine which file formats can be used with the scripture range than the controller.
___
*[`src/Serval.Translation/Services/EngineService.cs` line 175 at r7](https://reviewable.io/reviews/sillsdev/serval/289#-Nq-e6WY-Zg1fNJc_fqr:-Nq-e6WY-Zg1fNJc_fqs:b-d7hi8u) ([raw file](https://github.com/sillsdev/serval/blob/59aebce7036937f587a85908f958227cf596f869/src/Serval.Translation/Services/EngineService.cs#L175)):*
> ```Smalltalk
>             Dictionary<string, TrainingCorpus>? trainOn = build.TrainOn?.ToDictionary(c => c.CorpusRef);
>             var client = _grpcClientFactory.CreateClient<TranslationEngineApi.TranslationEngineApiClient>(engine.Type);
>             Dictionary<string, List<int>> GetChapters(V1.Corpus corpus, string scriptureRange)
> ```

You could use the `ScriptureRangeParser.GetChapters` convenience method instead.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 976 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-NpzhyEN9USLLgYPaIAm:-Nq-v8HJE7i05PFcA7qb:b-n5vl1v) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L976)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This exception is thrown if both the text ids and scripture ranges are specified.
</blockquote></details>

Ah - I stand corrected.


<!-- Sent from Reviewable.io -->

johnml1135 commented:
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 976 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-NpzhyEN9USLLgYPaIAm:-Nq-vEW05UpW2wZY07pN:b-6x9iou) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L976)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Ah - I stand corrected.
</blockquote></details>

Can the comment be updated to reflect the logic - that the user should `Set at most one of TextIds and ScriptureRange.`


<!-- Sent from Reviewable.io -->

Enkidu93 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: 11 of 13 files reviewed, 3 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 976 at r6](https://reviewable.io/reviews/sillsdev/serval/289#-NpzhyEN9USLLgYPaIAm:-Nq-vznUBWljZKTTLlp5:b-896fix) ([raw file](https://github.com/sillsdev/serval/blob/c22851907b262a92720a34277ba3a61d93605a4d/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L976)):*
<details><summary><i>Previously, johnml1135 (John Lambert) wrote…</i></summary><blockquote>

Can the comment be updated to reflect the logic - that the user should `Set at most one of TextIds and ScriptureRange.`
</blockquote></details>

Done.
___
*[`src/Serval.Translation/Controllers/TranslationEnginesController.cs` line 985 at r7](https://reviewable.io/reviews/sillsdev/serval/289#-Nq-dVUQDLBh4XGqOja1:-Nq-wLS--16ulYNZisEM:bb6un1j) ([raw file](https://github.com/sillsdev/serval/blob/59aebce7036937f587a85908f958227cf596f869/src/Serval.Translation/Controllers/TranslationEnginesController.cs#L985)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This check should be performed in the `EngineService`. I think I would rather the engine service determine which file formats can be used with the scripture range than the controller.
</blockquote></details>

Done
___
*[`src/Serval.Translation/Services/EngineService.cs` line 175 at r7](https://reviewable.io/reviews/sillsdev/serval/289#-Nq-e6WY-Zg1fNJc_fqr:-Nq-vxK2DOJp8lZGlgG_:b-896fix) ([raw file](https://github.com/sillsdev/serval/blob/59aebce7036937f587a85908f958227cf596f869/src/Serval.Translation/Services/EngineService.cs#L175)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

You could use the `ScriptureRangeParser.GetChapters` convenience method instead.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

johnml1135 reviewed (approved):
> Reviewed 1 of 3 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289)* status: all files reviewed, 1 unresolved discussion (waiting on @ddaspit)



<!-- Sent from Reviewable.io -->

ddaspit reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/serval/289#-:-Nq2xXqsDBCepdTQFilN:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Enkidu93)



<!-- Sent from Reviewable.io -->
